syntax = "proto3";

package chat;

service ChatService {
  rpc CreateAccount(CreateAccountRequest) returns (CreateAccountResponse);
  rpc Login(LoginRequest) returns (LoginResponse);
  rpc DeleteAccount(DeleteAccountRequest) returns (BasicResponse);
  rpc GetUsers(GetUsersRequest) returns (GetUsersResponse);
  rpc SendMessage(SendMessageRequest) returns (SendMessageResponse);
  rpc SearchUsers(SearchUsersRequest) returns (SearchUsersResponse);
  rpc GetRecentChats(GetRecentChatsRequest) returns (GetRecentChatsResponse);
  rpc GetPreviousMessages(GetPreviousMessagesRequest) returns (GetPreviousMessagesResponse);
  rpc GetChatUnreadCount(GetChatUnreadCountRequest) returns (GetChatUnreadCountResponse);
  rpc GetUnreadMessages(GetUnreadMessagesRequest) returns (GetUnreadMessagesResponse);
  rpc DeleteMessages(DeleteMessagesRequest) returns (BasicResponse);
}

// Basic response for simple operations
message BasicResponse {
  int32 code = 1;
  string message = 2;
}

// Account Management
message CreateAccountRequest {
  string email = 1;
  string username = 2;
  string password = 3;
}

message CreateAccountResponse {
  int32 code = 1;
  string message = 2;
}

message LoginRequest {
  string email = 1;
  string password = 2;
}

message UserData {
  string id = 1;
  string username = 2;
  string email = 3;
}

message LoginResponse {
  int32 code = 1;
  string message = 2;
  UserData user = 3;
}

message DeleteAccountRequest {
  string email = 1;
  string password = 2;
}

// User Management
message GetUsersRequest {
  int32 page = 1;
}

message GetUsersResponse {
  int32 code = 1;
  string message = 2;
  repeated UserData users = 3;
}

message SearchUsersRequest {
  string pattern = 1;
  int32 page = 2;
  string current_user_id = 3;
}

message SearchUsersResponse {
  int32 code = 1;
  string message = 2;
  repeated UserData users = 3;
  int32 total_pages = 4;
}

// Messaging
message SendMessageRequest {
  string content = 1;
  string recipient_id = 2;
  string sender_id = 3;
}

message MessageData {
  string message_id = 1;
  string sender_id = 2;
  string recipient_id = 3;
  string content = 4;
  string timestamp = 5;
}

message SendMessageResponse {
  int32 code = 1;
  string message = 2;
  MessageData data = 3;
}

message DeleteMessagesRequest {
  repeated string message_ids = 1;
}

// Chat Management
message GetRecentChatsRequest {
  string user_id = 1;
  int32 page = 2;
}

message LastMessage {
  string content = 1;
  string timestamp = 2;
  bool is_from_me = 3;
}

message ChatData {
  string user_id = 1;
  string username = 2;
  int32 unread_count = 3;
  LastMessage last_message = 4;
}

message GetRecentChatsResponse {
  int32 code = 1;
  string message = 2;
  repeated ChatData chats = 3;
  int32 total_pages = 4;
}

message GetPreviousMessagesRequest {
  string user_id = 1;
  string other_user_id = 2;
  int32 page = 3;
}

message MessageSender {
  string user_id = 1;
  string username = 2;
}

message ChatMessage {
  string message_id = 1;
  string content = 2;
  string timestamp = 3;
  bool is_from_me = 4;
  MessageSender sender = 5;
}

message GetPreviousMessagesResponse {
  int32 code = 1;
  string message = 2;
  string user_id = 3;
  string other_user_id = 4;
  repeated ChatMessage messages = 5;
  int32 total_pages = 6;
}

message GetChatUnreadCountRequest {
  string user_id = 1;
  string other_user_id = 2;
}

message GetChatUnreadCountResponse {
  int32 code = 1;
  string message = 2;
  string user_id = 3;
  string other_user_id = 4;
  int32 count = 5;
}

message GetUnreadMessagesRequest {
  string user_id = 1;
  string other_user_id = 2;
  int32 num_messages = 3;
}

message UnreadMessage {
  string message_id = 1;
  string sender_id = 2;
  string recipient_id = 3;
  string content = 4;
  string timestamp = 5;
  bool is_read = 6;
  bool is_from_me = 7;
}

message GetUnreadMessagesResponse {
  int32 code = 1;
  string message = 2;
  repeated UnreadMessage messages = 3;
}
